crates/turborepo-lib/src/graph/walker.rs

```diff
@@ -0,0 +1,257 @@
+use std::{collections::HashMap, hash::Hash};
+
+use futures::future::join_all;
+use petgraph::{
+    visit::{IntoNeighborsDirected, IntoNodeIdentifiers},
+    Direction,
+};
+use tokio::{
+    sync::{broadcast, mpsc, oneshot, watch},
+    task::JoinHandle,
+};
+use tracing::log::trace;
+
+pub struct Walker<N> {
+    cancel: watch::Sender<bool>,
+    node_events: Option<mpsc::Receiver<(N, oneshot::Sender<()>)>>,
+    join_handles: Vec<JoinHandle<()>>,
+}
+
+pub type WalkMessage<N> = (N, oneshot::Sender<()>);
```

Comment:
We could opt for making smart pointer that automatically sends the done signal upon dropping (think `MutexGuard`), but I opted to be explicit to avoid accidental done signals. I think it would be too easy to accidentally drop the guard type since the inner type implements `Copy` and any caller will probably want to resolve the node index to the node weight for processing.